### DISCUSSION TITLE
Making the github-action work with a time-limited AWS CodeArtifact authorization token (using env variable for the token)

### DISCUSSION THREAD

**Discussion by brybicki-cyferd** [renovatebot/renovate#11284] [Request Help]
Hi there,

I'm trying to get Renovate to work with a private Npm registry hosted in AWS CodeArtifact. Even though our repositories are hosted on GitHub I had to dismiss the Renovate GitHub App since according to this https://docs.renovatebot.com/getting-started/private-packages/ the preferred method of authentication in my scenario is:

> 3. Renovate App with a private registry: Add an unencrypted npmrc plus an encrypted npmToken in config

As you are probably already aware CodeArtifact limits the lifespan of authorization tokens to 12 hours. When using the above auth method the token is supposed to be encrypted using a GUI web app which uses Renovate App’s public key as the encryption key. This is a problem since we cannot replicate this step programmatically (to get over the 12h token lifespan problem).

That brought me to the GitHub Action - https://github.com/renovatebot/github-action. I have used this discussion https://github.com/renovatebot/renovate/discussions/8033 along with the repo's README as the basis for my solution:

`.github/renovate.js`
```
module.exports = {
  branchPrefix: 'test-renovate/',
  dryRun: false,
  gitAuthor: 'Renovate Bot <renovate@whitesourcesoftware.com>',
  onboarding: true,
  platform: 'github',
  includeForks: true,
  repositories: [
    'myorg/myrepo',
  ],
  packageRules: [
    {
      matchDatasources: ['npm'],
      registryUrls: ['https://myorg.d.codeartifact.eu-central-1.amazonaws.com/npm/my_priv_registry'],
      description: 'lockFileMaintenance',
      stabilityDays: 0,
      dependencyDashboardApproval: false,
      matchUpdateTypes: [
        'pin',
        'digest',
        'patch',
        'minor',
        'lockFileMaintenance',
      ]
    }
  ],
  hostRules: [
    {
      hostType: 'npm',
      matchHost: 'https://myorg.d.codeartifact.eu-central-1.amazonaws.com/npm/my_priv_registry/',
      token: process.env.CODEARTIFACT_AUTH_TOKEN,
    },
  ],
};
```
I've made some adjustments from what was suggested in the discussion since the config validator kept saying I needed to migrate my config (I'm guessing the syntax has changed since). The key line is `token: process.env.CODEARTIFACT_AUTH_TOKEN` which I thought would allow me to dynamically set the token at action runtime. I can pass AWS CLI creds as secrets to the GitHub action runner, query the API to grab my token, and put it in an env variable like so:

`.github/workflows/renovate.yml`
```
name: Renovate
on:
  schedule:
    # The "*" (#42, asterisk) character has special semantics in YAML, so this
    # string has to be quoted.
    - cron: '0/10 * * * *'
jobs:
  renovate:
    runs-on: ubuntu-latest
    env:
      AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
      AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
    steps:
      - name: Set CA token env var
        run: export CODEARTIFACT_AUTH_TOKEN=$(aws codeartifact get-authorization-token --domain myorg \
             --domain-owner myorgid --region=eu-central-1 --query authorizationToken --output text )

      - name: Checkout
        uses: actions/checkout@v2.0.0

      - name: Renovate Action
        uses: renovatebot/github-action@v25.72.2
        with:
          configurationFile: .github/renovate.js
          token: ${{ secrets.RENOVATE_GH_TOKEN }}
```
Although I do wonder if steps `Set CA token env var` and `Renovate Action` use the same environment or if the env variable is out of scope once execution gets to `Renovate Action`.

At first I was running this with `onboarding: false`  but the Renovate Action didn't open any pull-requests and printed this:
```
INFO: Repository started (repository=myorg/myrepo)
       "renovateVersion": "25.72.2"
 INFO: Repository is disabled - skipping (repository=myorg/myrepo)
```
I figured the repo needed to be properly onboarded after all and made the change to `onboarding: true`. At this point the configure PR was opened and "What to Expect" section looked good. However Renovate didn't manage to resolve my private dependencies and all of them came up under "Dependency Lookup Warnings".
```
Please correct - or verify that you can safely ignore - these lookup failures before you merge this PR.

Failed to look up dependency @myorg/myrepo
etc...
```

When I checked the action's log I found these lines which explain it:
```
 INFO: Repository started (repository=myorg/myrepo)
       "renovateVersion": "25.72.2"
 INFO: Fallback to renovate.json file as a preset is deprecated, please use a default.json file instead. (repository=myorg/myrepo)
...
WARN: Custom registries are not allowed for this datasource and will be ignored (repository=myorg/myrepo)
       "datasource": "npm",
       "registryUrls": [
         "https://myorg.d.codeartifact.eu-central-1.amazonaws.com/npm/my_priv_registry"
       ]
```
So as far as I can see my `renovate.js` never got used since it seems to have been deprecated and custom registries are not allowed for the npm datasource. Although I must say I don't fully understand this `Fallback to renovate.json file as a preset is deprecated` since I wasn't aware I was attempting to use a preset.

Questions:

1. Can the renovate github-action only support JSON configs now or am I doing something wrong with my `renovate.js`?
2. If a `.js` config is no longer supported, are there any other ways to refresh the token programmatically? I'm wondering if I can re-write a JSON config on the runner after checkout and place a plain-text token there. But then again:
3. I was supplying the token in `hostRules` on `matchHost` which doesn't seem to work for "custom registries". Am I doing something that's not permitted in this block? Any alternatives?
4. If going the GitHub Action route is not viable could I make this work using either the docker image or a node.js server? How would I do that (generally)? Or will I run into the same issue of not being able to set a custom npm registry and/or refresh the token?

This thread also looks related - https://github.com/renovatebot/renovate/issues/11207, but its author didn't include his configs so I'm not sure of his exact setup.

Any help would be greatly appreciated. That time-limited CodeArtifact token is being the bane of my life right now.

**Comment by rarkins** [comment]
1. You use a `config.js`/`config.json` file in the current working directory where you run the bot (optionally - CLI or env are also supported), but need to have a `renovate.json` or equivalent in the destination repo. If you are running in the same one you could set onboarding=false and requireConfig=false together so that it doesn't skip over, and then you don't need a `renovate.json`.
2. JS is supported, I wonder if maybe you were just missing requireConfig=false to skip a `renovate.json`. I'm not certain you can execute the functions you need in the `config.js` though, because we "require" it.
3. I don't understand this question
4. I think we should aim to support the rotating Code Artifact tokens natively in Renovate so that it works in the app too. It would be good if you can create a feature request for this with adequate details if none exists already. We support somethign similar for ECR already

**Reply by brybicki-cyferd** [reply]
Thanks for getting back to me so quickly, it's appreciated. I'm doing some further testing after reading your explanations and will get back with a further update tomorrow.

So far it looks like you were right about this:
> I wonder if maybe you were just missing requireConfig=false to skip a renovate.json.

After adding that in and setting `onboarding=false` again, Renovate opened version bumping PRs immediately without explicit onboarding of my repository.

> 2. I'm not certain you can execute the functions you need in the config.js though, because we "require" it.

I'm not sure I understand what you mean. If JS is supported then the only thing that I really need working in my `renovate.js` is setting the token from an environment variable which I'm trying to achieve with this `token: process.env.CODEARTIFACT_AUTH_TOKEN,`. Well this and setting a private registry URL.

> 3. I don't understand this question

Apologies, that didn't make sense because I thought this `WARN` was thrown when processing `hostRules[]` when it actually came from processing the `packageRules[]`.
```
packageRules: [
    {
      matchDatasources: ['npm'],
      registryUrls: ['https://cyferd-893024013534.d.codeartifact.eu-central-1.amazonaws.com/npm/cyferd-common'],
...
```

```
WARN: Custom registries are not allowed for this datasource and will be ignored (repository=Cyferd/renovate-poc)
       "datasource": "npm",
       "registryUrls": [
         "https://cyferd-893024013534.d.codeartifact.eu-central-1.amazonaws.com/npm/cyferd-common"
       ]
```
I'm basically trying to follow your instructions from this old discussion https://github.com/renovatebot/renovate/discussions/8033#discussioncomment-214374 and don't understand why that warning is being shown. I've changed `"datasources": ["npm"],` to `matchDatasources: ['npm'],` as the config validator suggested it. Just like the OP in that discussion I'm also trying to get Renovate to only use the private CodeArtifcat registry. Our CodeArtifact domain contains a sort of a mirror connected to the default public npm registry (https://registry.npmjs.com/) which can be used if the package is not found in the private one. I know this isn't ideal but it's like things got set up before I showed up. Normally I'd only use the private registry for our private packages but some of them have not been properly scoped and I didn't have a chance to track them down and fix it yet.

**Reply by rarkins** [reply]
You need to use a `.npmrc` `registry=` line to control the registry URL which is used.

**Comment by viceice** [comment]
The renovate action is using docker to run renovate, so not all env vars are passed.

https://github.com/renovatebot/github-action/blob/91d387da5a1e0baeff7f3539903cfe8acd286e9b/src/input.ts#L11

So you need to configure hostRules configuration using environment variables
https://docs.renovatebot.com/getting-started/private-packages/#hostrules-configuration-using-environment-variables

**Reply by brybicki-cyferd** [reply]
I realized I was relying on the onboarding PR to tell me if all dependencies got resolved. Having a bit of a trouble testing this without it. I'm not sure if behaviour I'm expecting is actually how things should work.

Let's say I have a project with both public and private dependencies, e.g.:
```
"dependencies": {
    "name": "renovate-poc",
    "@myorg/common-api": "1.0.11",
    "lodash": "4.17.20"
```
Then, let's also assume that my private project `@myorg/common-api@1.0.11` lists a vulnerable public dependency in its `package.json`. I fix that and release `@myorg/common-api@1.0.12` which is now vuln-free.

If I set things up correctly, should I expect Renovate to discover this on the next run and open a PR in the `renovate-poc` repo to bump `@myorg/common-api` from `1.0.11` to `1.0.12`? Or will it only ever bump up public packages?

**Reply by rarkins** [reply]
It will bump any packages that it can see/look up. When using a private registry it doesn't even really understand the difference between public and private at times

**Reply by brybicki-cyferd** [reply]
> The renovate action is using docker to run renovate, so not all env vars are passed.

Thanks for explaining this. I decided to remove the additional complexity of dealing with docker (with using the Renovate Action) and use the Renovate CLI npm tool directly on a GitHub Actions Runner. I found it easier to pass the required vars and configure my `hostRules` this way.

I can now use this in my `.github/renovate.js`:
```
hostRules: [
    {
      hostType: 'npm',
      matchHost: 'https://myorgID.d.codeartifact.eu-central-1.amazonaws.com/npm/myorg-common/',
      token: process.env.CODEARTIFACT_AUTH_TOKEN,
    },
```
along with this Action step:
```
- name: Auth and Run Renovate CLI
        run: |
          export CODEARTIFACT_AUTH_TOKEN=$(aws codeartifact get-authorization-token ...)
          renovate
```

I now have a semi-working solution and Renovate opens PRs to bump my private packages. Although for some reason it fails to generate required package-lock.json changes due to problems with authentication. It's strange since I know that authentication to do the package lookup was already successful because Renovate noticed a new version in my private registry.

```
WARN: artifactErrors (repository=myorg/renovate-poc, branch=test-renovate/myorg-renovate-poc-sample-project-1.x)
       "artifactErrors": [
         {
           "lockFile": "package-lock.json",
           "stderr": "npm WARN read-shrinkwrap This version of npm is compatible with lockfileVersion@1,
            but package-lock.json was generated for lockfileVersion@2. I'll try to do my best with it!
           \nnpm ERR! code E401\nnpm ERR! Unable to authenticate, need: ***"myorg/myorg-common\",
           Basic realm=\"myorg/myorg-common\"\n\nnpm ERR! A complete log of this run can be found in:
           ...
```
It makes more sense after reading how lock files are handled (being handed off to the package manager itself) - https://docs.renovatebot.com/getting-started/use-cases/:

1. The repository has a `package.json` and `package-lock.json` containing version `1.0.0` of a dependency
2. Renovate determines that version `1.1.0` is available
3. Renovate patches the `package.json` to change the dependency's version from `1.0.0` to `1.1.0`
4. Renovate then runs `npm install`, which triggers `npm` to update the `package-lock.json` accordingly
5. Renovate then commits both the `package.json` and `package-lock.json` files together for the PR

So right now steps 1 to 3 are working with the auth I already set up but `npm install` in step 4 fails since `npm` on its own doesn't have my token in its context. I'll update this thread with all my configs if I manage to get over this problem too.

**Reply by rarkins** [reply]
The error indicates that you running npm v6 with Renovate while your project's lock file has been generated with npm v7. i.e. unrelated to authentication. Update: actually that was just a warning. the error was related to auth

**Reply by brybicki-cyferd** [reply]
Renovate gave me more details in a PR comment I just didn't mention it:
```
npm WARN read-shrinkwrap This version of npm is compatible with lockfileVersion@1, but package-lock.json was generated for lockfileVersion@2. I'll try to do my best with it!
npm ERR! code E401
npm ERR! Unable to authenticate, need: Bearer realm="myorg/myorg-common", Basic realm="myorg/myorg-common"
```
I should maybe update npm on the GH runner before running `renovate` but this is only generating a warning. The error is thrown on the auth step

**Reply by brybicki-cyferd** [reply]
I was able to get over the last issue by also running `aws codeartifact login` in my `Auth and Run Renovate CLI` step. That command updates the local default `.npmrc` on disk with a private registry link and a new hard-coded token. According to the docs Renovate itself is not able to use an npmrc from disk:

https://docs.renovatebot.com/getting-started/private-packages/#private-config-presets
> However if you do still use them, private modules should work if you configure the npmrc file including token credentials in your bot admin config. Credentials stored on disk (e.g. in ~/.npmrc) are no longer supported.

So it most likely wouldn't work for step `2.` of a Renovate run (that's why I still use another token in an env var) but it seems to work for step `4.` when renovate runs `npm install`. This seems to be independent of Renovate itself so npm functions as normal and firstly reads the local `.npmrc`.

Using 2 tokens is definitely not pretty but this setup seems to work. I'll do some more testing and see if I can polish this solution off a bit then write an update.

**Reply by rarkins** [reply]
if you push the token to a host rule in the right format then Renovate will add it to any `.npmrc` or create a new `.npmrc` before calling `npm`.

**Reply by brybicki-cyferd** [reply]
Could you point me to the relevant piece of code please? I'm wondering what exactly gets added to a `.npmrc` by Renovate. Things didn't work for me before I added the `aws codeartifact login` command, but I'm not sure if the token wasn't passed from my `hostRules` or if it only appeared like it wasn't because the registry wasn't set to my private one in that `.npmrc`.

Although I followed your advice:

> You need to use a .npmrc registry= line to control the registry URL which is used.

and added this line to `renovate.json`:
```
"npmrc": "@myorg:registry=https://myorgid.d.codeartifact.eu-central-1.amazonaws.com/npm/myorg-common/"
```

Should that also have been passed to the `.npmrc` file along with the token from `hostRules` before Renovate ran `npm install`? Or does this stay "internal" only within the scope of Renovate (i.e. works only for the package version check stage but isn't used for the `npm install` command)?

**Reply by viceice** [reply]
you need that npmrc config, which should better be in repo `.npmrc`.

`npmrc` config option works too and will fully overwrite repo `.npmrc`

You need the hostrule you posetd above too, so renovate will add an auth line to the repo `npmrc` before running npm install.

Here is how we generate additional auth lines to repo `.npmrc` file
https://github.com/renovatebot/renovate/blob/714e51a316b5909219c097052c99148d7a1b7046/lib/manager/npm/post-update/rules.ts#L10-L49

**Reply by viceice** [reply]
To pass the generated token, you should use something like this:

```yaml
name: Renovate
on:
  schedule:
    # The "*" (#42, asterisk) character has special semantics in YAML, so this
    # string has to be quoted.
    - cron: '0/10 * * * *'
jobs:
  renovate:
    runs-on: ubuntu-latest
    env:
      AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
      AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
    steps:
      - name: Set CA token env var
        run: |
          export CODEARTIFACT_AUTH_TOKEN=$(aws codeartifact get-authorization-token --domain myorg \
             --domain-owner myorgid --region=eu-central-1 --query authorizationToken --output text )
          echo "RENOVATE_SECRETS='{\"CODEARTIFACT_AUTH_TOKEN\": \"${CODEARTIFACT_AUTH_TOKEN}\"}'" >> $GITHUB_ENV

      - name: Checkout
        uses: actions/checkout@v2.0.0

      - name: Renovate Action
        uses: renovatebot/github-action@v26.4.2
        with:
          configurationFile: .github/renovate.js
          token: ${{ secrets.RENOVATE_GH_TOKEN }}
```

**config.js**
```js
module.exports = {
  branchPrefix: 'test-renovate/',
  dryRun: false,
  gitAuthor: 'Renovate Bot <renovate@whitesourcesoftware.com>',
  onboarding: true,
  platform: 'github',
  includeForks: true,
  repositories: [
    'myorg/myrepo',
  ],
  packageRules: [
    {
      matchDatasources: ['npm'],
      description: 'lockFileMaintenance',
      stabilityDays: 0,
      dependencyDashboardApproval: false,
      matchUpdateTypes: [
        'pin',
        'digest',
        'patch',
        'minor',
        'lockFileMaintenance',
      ]
    }
  ],
  hostRules: [
    {
      hostType: 'npm',
      matchHost: 'https://myorg.d.codeartifact.eu-central-1.amazonaws.com/npm/my_priv_registry/',
      token: '{{ secrets.CODEARTIFACT_AUTH_TOKEN }}',
    },
  ],
};
```

**renovate.json**
```json
{
  "npmrc": "@myorg:registry=https://myorg.d.codeartifact.eu-central-1.amazonaws.com/npm/my_priv_registry/"
}
```

**Reply by brybicki-cyferd** [reply]
Thank you for your help guys.

The main thing I was missing was `always-auth=true` in the nprmc lines I was passing. Additionally I made my private registry scoped since then so now the nprmc line in `config.js` looks like this:
```
npmrc: 'registry=https://registry.npmjs.com/\n@myorg:registry=https://myorg-myid.d.codeartifact.eu-central-1.amazonaws.com/npm/myorg-common/\n//myorg-myid.d.codeartifact.eu-central-1.amazonaws.com/npm/myorg-common/:always-auth=true',
```
In addition to this host rule:
```
hostRules: [
    {
      hostType: 'npm',
      matchHost: 'https://myorg-myid.d.codeartifact.eu-central-1.amazonaws.com/npm/myorg-common/',
      token: process.env.RENOVATE_CA_TOKEN
    },
```

This allows me to programmatically refresh the `RENOVATE_CA_TOKEN` variable and pass it to Renovate. I no longer need two tokens to make it work.

As a side note for people who'll be dealing with tens or hundreds of repositories that run their CI on a GH Actions runner - you will not fit this into your allocated GH Action's minutes unless you're on Enterprise tier. Renovate does A LOT of branch rebasing and each rebase triggers a CI run. After merging one PR to bump a version (into master) Renovate will need to rebase the next version bumping PR, which will trigger CI yet again, and so on and so on... Even with CI on Jenkins it took Renovate about 12 minutes to go through 26 repos when using the Renovate GH Action (with 3 PRs per repo as a hard limit). If you run it every hour (as recommended) that's already a lot of minutes even without CI on GH Actions. I recommend looking into self-hosting unless this is meant for your hobby project.

**Comment by rarkins** [comment]
Related: #11325
